### ISSUE TITLE
storage: allow low-level filtering in `Scan`

### ISSUE THREAD

**Issue by tbg** [cockroachdb/cockroach#20080]
In the upcoming PR to optimize the GC queue (#20052) one interesting item is
being able to issue a Scan that does not return any results. This seems easy to
implement (though its users would have to accept that in mixed clusters, results
may still be returned even if not desired).

This is somehow a very special case of a general problem, namely applying a
function to the results to determine the result set (i.e. morally sending a
filter closure along with the `Scan`)

For example, picture the following (simplified) protos (tag 4 is half of
https://github.com/cockroachdb/cockroach/issues/20059):

```proto
enum ScanOpType {
  OP_ERROR = 0;
  OP_COUNT = 1; // return single result ("", <count>)
  OP_MATCHPREFIX = 2; // return kv only if key matches prefix
  OP_DISCARD = 3; // just drop everything (the GC queue use case)
  OP_STRIPPREFIX = 4; // strip x characters from returned keys
  // ...
}
```

```proto
message ScanRequest {
  Span header = 1;
  ScanOpType op = 2;
  repeated bytes op_args = 3;
}
```

A node asked to evaluate such a `ScanRequest` looks up the implementation from
the enum (error'ing if it's not supported or the arg count doesn't match) and
passes this down to the MVCC layer, which then passes the read key-value pairs
through the implementation. As a bonus, keys can be inspected at a very low
level, before being copied into Go-managed memory, making "picky" ops much
cheaper.

It seems straightforward enough to add new impls between releases (the migration
is straightforward), though we may likely never want to change existing impls.

I'm sure our SQL folks have thought about this already and perhaps even filed an
issue that I missed. Curious whether I'm missing something obvious.

(At the very least, what I propose above isn't expressive enough - for `COUNT(*)`, for example, you'll want to filter a subset of kv pairs (one for each row) and then return only a count, suggesting that instead a slice of ops should be specified and those are then chained. And combining batches needs to know how to aggregate partial results).

cc @RaduBerinde

**Comment by nvanbenschoten** [comment]
> At the very least, what I propose above isn't expressive enough

Maybe we should just add a full SQL query language to `ScanRequests` :trollface: 

Seriously though, @andreimatei and I have discussed this before a few times (which I assume means @andreimatei and most people have discussed this a few times). From a SQL perspective, the way we had pictured this being most useful was to expose something kind of like your `OP_MATCHPREFIX`, but more general. What we really want is some kind of pattern for keys to match against. Only keys matching the pattern would be returned.

There are a number of cases where this is exactly what SQL needs to minimize the amount of unnecessary data retrieved. For instance:
- Scanning a table for only certain columns/families. Imagine a table has a number of very large columns and a few small ones. For simplicity, pretend their each in their own column family. Currently, even if a query only needs a single column, it will need to send a `ScanRequest` over the entire table, fetching all of the large columns as well. These large columns will then be pulled out of RocksDB, shipped over the network, decoded, and then thrown away immediately. Very wasteful! 
Ideally, SQL would be able to specify a pattern like `<TableID>/1/*/6` where `1` is the primary index, the wildcard is on the primary key columns, and `6` is the column family of the only desired column. This is something that came up in https://github.com/cockroachlabs/rubrik/issues/5, and the only way I could think of to address it there was to create a separate index covering only the desired columns for that query so that we could avoid the primary index altogether. Needing to do that seems like an action dictated by a deficiency in our kv API.
- Similarly, I think scanning an interleaved parent or child table in isolation will run into the same issue. Right now I believe we collect and return the KVs for the parent and all child tables, even if only one is desired, before dropping the unneeded KVs on the floor at the SQL layer. Ideally, we'd be able to specify something like:
```
scan parent pattern: <parentID>/1/*/[^#]
scan child1 pattern: <parentID>/1/*/#/<child1ID>/1
scan child2 pattern: <parentID>/1/*/#/<child2ID>/1

(where # is the interleave sentinel)
```

**Comment by andreimatei** [comment]
I don't want to troll more than @nvanbenschoten , but I think this is the place for one of @knz's stack machines (or BPF or, dare I say, native code) - I think we should pass programs to the scan rather than an ever-growing set of options.
We can start by writing some C++ functions by hand for the suggested use cases and have them be primitives in our assembly.

**Comment by tschottdorf** [comment]
> I don't want to troll more than @nvanbenschoten , but I think this is the
place for one of @knz's stack machines (or BPF or, dare I say, native code) - I
think we should pass programs to the scan rather than an ever-growing set of
options.

Don't think that is trolling, that's kind of the holy grail at the end
(especially when the resulting code gets optimized by whichever JIT compiler
translates it), but obviously that'll need a lot more juice.

For today, I ran an nefarious yet interesting hack, namely not allocating any
key/value memory in Go-land for a `COUNT(*)` on `tpch.lineitem` (29999795 rows)
on a GCE 3node cluster.

<details>
<summary>TL;DR: in the particular case below, shaves `>20%` off.</summary>

- set up a 3x ephemeral GCE cluster (`n1-standard-4`, ssd)
- 
  ```
  CREATE DATABASE tpch;
  RESTORE tpch.lineitem FROM 'azure://backup-test/benchmarks/tpch/scalefactor-5?...';
  --------------------+-----------+--------------------+----------+---------------+----------------+------------+
  |       job_id       |  status   | fraction_completed |   rows   | index_entries | system_records |   bytes    |
  +--------------------+-----------+--------------------+----------+---------------+----------------+------------+
  | 297793942740172801 | succeeded |                  1 | 29999795 |     239998360 |              0 | 6960275076 |
  +--------------------+-----------+--------------------+----------+---------------+----------------+------------+
  (1 row)
  
  Time: 15m13.506980945s
  ```
- repeatedly run the below on master (1c681373b5b4b973a1f3c21a91899b963cb06f9b)
  ```
  select count(*) from tpch.lineitem;
  +----------+
  | count(*) |
  +----------+
  | 29999795 |
  +----------+
  (1 row)
  
  Time: 30.570480348s // best one I've seen, usually 32-33s
  ```
- hack the binary so that the ScanRequests return `nil` `Key` and `RawBytes` slices
- additionally hack the MVCC code so that the key and bytes are never copied over from C++-land (never use the `BytesAlloc`)
- re-run the queries repeatedly:
  ```
  root@:26257/> select count(*) from tpch.lineitem;
  +----------+
  | count(*) |
  +----------+
  | 29999795 |
  +----------+
  (1 row)
  
  Time: 24.21286419s // pretty consistently
- re-run on master to verify timings are about the same (yes)

So it looks that for that particular setup, we shave 6+ seconds off the workload (or `>20%`).

</details>

<br />

The example here is a little contrived, but to be fair a `COUNT(*)` is a fairly common query
and we ought to optimize it. Of course what I did is way too blunt as it only applies to the
case of one column family, no interleaved tables, ...

This approach here (filtering what we pull over into Go) is somewhat
complementary to #19288, in which @arjunravinarayan and @asubiotto are looking
at how fast we can generally pull data off the disk (and are finding that the
cgo crossings are too expensive, so it's better to push much of the iteration
into C++). I suspect that the savings we see here are to some extent that we
allocate less (the `ByteAllocator` allocates in the teens, so the overhead
doesn't seem terrible but noticeable) but also that GC has to keep track of less
pointers (all `.Key` and `.RawValue` are `nil`).

I think the 95%-gains goal is something like this (the first two being what I
think is considered in #19288):

- Iteration happens in C++ (i.e. we get to throw data into Go as fast as we possibly can in the absence of optimizing RocksDB)
- the result of the Scan in Go is represented as an opaque blob (i.e. GC sees one pointer)

and then, to limit the amount of data we have to throw into Go,

- Scans pass a compact byte slice "program" (to C++) that is simple but expressive enough for the typical filters, allowing scans to omit rows the kvFetcher doesn't care about.
  I *think* (?) that these are all questions of the form

  > is this for the index and column family I'm scanning or some other column or interleaved table and can I pick the values I really want?

  and can thus be expressed as a sequence of operations like

  - skip the next encoded value
  - copy the next encoded value into the output
  - match the byte string X against the current position in the byte array (on failure, kv pair is discarded)

  (with different programs applied to the key and the value)
  
  This would also naturally enable "prefix compression" where scans don't return a shared prefix (though the program would do that work).
  Would need to make this work with `ResumeSpan`, though. Can be avoided if we remove the "copy into output" instruction and make it a pure
  emit/skip decision, for now. cc @jordanlewis since this appears to subsume https://github.com/cockroachdb/cockroach/issues/20059.

The loftier "compiled programs" would allow checking predicates (i.e. is the
encoded int at the n'th position >=5?) and potentially run as code that can make
use of vectorization, which would be useful for highly selective table scans, but
I don't think there's a very good reason to dip into that pool right away.

Code at https://github.com/tschottdorf/cockroach/tree/experiments/countstar.

**Comment by tschottdorf** [comment]
I looked around what our options for just-in-time-compilation would be, and https://github.com/bitfunnel/nativejit/ looks pretty interesting.

- supports all platforms
- only depends on C++ runtime
- can call C functions
- generated code is optimized
- low latency (or so they say)

**Comment by arjunravinarayan** [comment]
Upon further investigation, NativeJIT appears to have no recent commit progress, with known bugs that will not be fixed. It is also unclear if LLVM is in fact slower, given its endless customizability (e.g. we could disable optimization passes that we can rule out since we have a fairly limited program domain).

I think we should proceed with LLVM for now, although I'm doing some additional research on alternatives.

Do note @tschottdorf that the 20% improvements you're seeing are a lower bound. From earlier experimentation (#19288) there are multiple bottlenecks in the COUNT(*) pipeline, and you need to solve both to get the full speedup. So it is somewhat likely that this improvement is greater than 20%, but you're not seeing the full speedup as the DistSQL aggregator processor is the new bottleneck.

**Comment by tschottdorf** [comment]
Some more reading:

- [JIT-Compiling SQL Queries in PostgreSQL Using LLVM](https://www.pgcon.org/2017/schedule/events/1092.en.html)
- https://www.slideshare.net/MemSQL/putting-compilers-to-work (notable because [memsql does he llvm thing too](http://blog.memsql.com/memsql-5-ships/))
- http://engineering.pivotal.io/post/codegen-gpdb-qx/

BTW, my branch is here:
 https://github.com/cockroachdb/cockroach/compare/master...tschottdorf:poc/llvm

At the time of writing, it does little more than bundle llvm and build a little bit of IR that shows that "everything we need to work works" (i.e. can just-in-time compile IR that calls external functions and call it, including from Go (that latter part hopefully not being necessary in the end)). After some of @arjunravinarayan's upcoming performance work to move scans below cgo and unpushed work of @andreimatei on a toy language we should be able to hack together an unholy POC.

**Comment by tschottdorf** [comment]
actually, @apavlo's [slides on query compilation from 15-721](http://15721.courses.cs.cmu.edu/spring2016/slides/19.pdf) are really helpful and have relevant bibliography.

**Comment by tschottdorf** [comment]
and http://highscalability.com/blog/2016/9/7/code-generation-the-inner-sanctum-of-database-performance.html has some detail on how this works in memsql.

**Comment by knz** [comment]
Folk it's not so clear to me that native code generation is the right way to go *at this time*.
The engineering expense is simply very large, and we'd have to create a new methodology to investigate problems.

What do you think about starting with a bytecode-based solution for now? With a C++ interpreter? There are several advantages:

- it sets up the scene for compilation later
- it postpones the discussion about portability of the solution
- it makes it easier to instrument the execution for debugging
- it sets up a baseline behavior to investigate compilation woes later, by making it possible to compare the bytecode interpretation to the native code behavior

To the argument "it's not as fast" I'll simply reply that "it's probably already orders of magnitude faster than what we have now".

**Comment by apavlo** [comment]
@knz is correct. Switching to a compilation based engine is a major undertaking. Using the LLVM makes it very difficult to debug (you don't get stack traces) and thus it makes difficult to find people to work on your system and contribute. You also need to build a (transactional) caching layer for compiled plans to minimize the compilation overhead.

You are free to use our codegen engine in the future as a starting point (Apache licensed): https://github.com/cmu-db/peloton/tree/master/src/codegen

We are also working on an LLVM IR interpreter (similar to https://github.com/andoma/vmir) for debugging.

**Comment by tschottdorf** [comment]
@knz @apavlo apologies, I should have been more clear in the intent from the beginning. My goal (personally) is not to actually try to build this beyond a prototype, but to dip my toes into LLVM (and codegen in general) and hopefully emerge with some demo elicits a "you did what?!" response (looks like that box is already checked 😄 )

Beyond that experiment, I'm definitely interested in a reasonable solution that we could eventually consider pushing past POC, and a bytecode interpreter (or even stack machine?) seems appropriate. Do you think there are existing options out there we may want to use, such as [Lua](https://www.lua.org/), or do we have to hand-roll our own?

**Comment by andreimatei** [comment]
I did not feel ready to have this conversations, but the conversation is upon us, so here's my current thoughts:
My personal interest in this area is 3-fold:
- do efficient filtering. This is based on a hunch that, for large scans with selective filters, a lot of time is spent directly and indirectly on a) filter evaluation and b) moving data around through the C++ -> Go interface.
- avoid a proliferation of scan-related APIs. If you look at how this present issue started, it was not about query compilation at all; it was about having different scan options that had nothing to do with SQL processing - something about processing intents.
- (personal) learn about generating and running native code, because it's cool. This objective has been achieved - I have created fire, so I'm happy.

Something I haven't paid specific though to is compilation of larger parts of SQL processing into some form of code. Perhaps some of @apavlo's "major undertaking" and debugging considerations may be alleviated when thinking about this generated code in more restricted contexts.

One thing that keeps resonating with me since I've learned about it is Linux kernel's use of BPF programs to provide very flexible APIs while keeping the API surface limited. So I kept thinking that ours scans should take something like that.

Now, when talking about compilation and code generation, I guess there are a couple of different, but intertwined, major questions:
1. What language should we express the programs we want to execute in? Should it be a high-level language (e.g. Lua,  C++, JS, a custom-built language) or a low-level language such as the LLVM IR?
2. What should we compile the programs too? Should it be native code or some assembly for a custom-built stack machine?
3. How do we execute the generated code? If it's native code, we might as well ask the CPU to execute it. If it's something else, should we build an interpreter, or find an already-built one.

I have spent some time over the past week playing with LLVM. I have to say, at the moment I'm very impressed with it. JIT-ing is specifically one of the things that it's built for (at least as of recent versions), and the API is much friendlier than I would have expected. Some bullet points:
a) the LLVM IR represents a nice level of coding - quite low, but still has things like types and functions. It's easy to work with it.
b) generating IR is easy; it has a bunch of facilities for inserting code at the right point, managing blocks of code, etc.
c) generating machine code out of the IR is trivial, and it seems to be quite portable. LLVM supports targeting a large set of platforms, and it deals with calling conventions (into external code) and other OS-specific crap for you.
d) what I had previously thought to be advanced and scary concepts just work easily. For example, it's really easy to control what symbols from the host process are available to the compiled modules, and LLVM does the linking/loading of these JITed modules for you.

In light of my good impression about all of the above, I think that directly or indirectly we should generate LLVM IR and compile it to machine code. I don't really see any disadvantages, only profit. What I don't yet have an opinion on is whether we should generate the IR directly, or use some other language. For expression evaluation, perhaps we could generate IR directly. But for other use cases (e.g. the intent processing thing), I think it might be useful to have a higher-level language that people can write code in directly.
Another option whispered around is Lua + LuaJIT. That I don't know about, but I'd probably take it over a home-grown solution.

I want to respond to a couple of of @knz's points in favor of a custom-built stack machine below.
But beyond the targeted responses, I think the biggest problem with that is that it'd be _more_ effort on our part. LLVM is quite E-Z.

So, the points were:

> it postpones the discussion about portability of the solution

I've become more confident in the portability of LLVM, so I think I'd welcome that discussion :)

> it makes it easier to instrument the execution for debugging

This I don't understand very well. What kind of instrumentation would you like that you can't have with native code? If we generated some higher-level code than we then JIT with LLVM, that code could have `printf()` calls.
@apavlo also mentioned stuff about debugging and about how you don't get stack traces from generated code. But what kind of useful stack traces would you get from an interpreter?

> it sets up a baseline behavior to investigate native compilation woes later, by making it possible to compare the bytecode interpretation to the native code behavior

Well but isn't this an argument for a lesser company than ours? We don't get promoted by improving the performance by 10% a quarte, so we don't have to throttle our improvements :)

> it's still compilation, so folk who like the idea will still have a toy to play with

This one doesn't speak to the alternatives.

**Comment by arjunravinarayan** [comment]
I strongly endorse @andreimatei's view.

**Comment by knz** [comment]
I'll be blunt here: 1) I fear you have no idea where you're getting into, and with regards to risk/payoff I believe we should as a team to dedicate our paid hours to projects that will more directly extend our runway and 2) I *currently* don't trust any of you guys to do training, aftercare and support durably after you ram a first working version in cockroachdb (you, and we as a team in general, have no/poor track record with propagating knowledge and maintenance skills on large sub projects), and 3) I certainly don't like the idea to multiply our code base by two with this project here at the same time we'll also multiply it by two for logical plan optimizations.

So tl;Dr, why not, but Friday time and schedule to release when we can afford to have 10+ engineers on the team that will be 100% dedicated to this layer (I.e in 3-4 years with current projections)
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by tschottdorf** [comment]
There are various dimensions on which opinions have been voiced (and for the record, I also feel fairly unauthoritative in my opinions, which no doubt reflects good judgment).

Regarding hand-rolled vs existing, I'm not sure there's a strong argument to hand-roll something made by anyone, so I assume we're all comfortable using something existing.

The next point of contention that has been muddled up into the LLVM-vs-alternatives debate is about JIT-vs-interpreted, and I'm pretty sure a convincing argument to prefer JIT'ed code has been made. On the other hand, interpreted code has a lot going for it: it's a lot faster than what we do now, and you don't have to worry about compilation overhead, which in turn avoids having to decide which queries you want to use it for -- you can just use it for everything.

The aforementioned https://github.com/andoma/vmir does interpret LLVM IR, but Andy pointed out that it's for debugging, and the README states that they'll focus on WASM in the future.

For that reason, I'm skeptical that LLVM is the right tool for the job. That said, if it had been decided that JIT'ing was the way to go, with my limited knowledge of the world out there I would also endorse LLVM with much of the enthusiasm @andreimatei has voiced as its JIT capabilities are seeing a lot of polish, many others are using it, and it's been pleasant to use. But I don't think we're in that business, and @knz's blunt reservations resonate strongly with me: we don't know what we're getting into.

**Comment by knz** [comment]
By the way, perhaps WASM is a good direction to look into:

- WASM is much cleaner/simpler than LLVM
- there's immense demand for WASM integrations in different software stacks; if there's not a WASM interpreter/compiler for Go in the near future, we'd attract *tons* of street creds if we provided one to the Go community
- it's also rather straightforward to interpret efficiently even if we don't JIT it

**Comment by knz** [comment]
FWIW I has imagined we'd ship distsql processors as standalone WASM interpreter threads talking to TCP sockets directly (to other nodes), completely outside of Go.

**Comment by andreimatei** [comment]
Rafa, I'm confused about what your stance is :)
Are you generally for or against code generation? If for, are you for interpreting it but against compiling it to native code? If so, do you care about the level of the code that we're interpreting?
If native code generation is the point where you get off the bus, please tell me again why that is exactly - how's it harder to debug or maintain or explain or whatever than interpreting something?

Btw, my thinking has never gone above a) generating code for expression evaluation and scan-like operations (e.g. scanning just a parent out of a parent-child interleaved hierarchy). Generating code for higher-level operations like full DistSQL processors, I've got no opinion on at the moment.

**Comment by knz** [comment]
I'm against:

- native code generation that early, 
- LLVM (its a very large piece of code, with many features that are not relevant to us)
- anything we can't start with by interpreting

I'm for:

- query compilation, predicate compilation, or whatever lowering of the logical query into some encoding that can be efficiently executed from our C++ layer (outside of Go).

As to "why it's difficult" basically I'm saying "you don't know what you're getting into". I barely scratched the surface of this kind of project myself over a duration of 6 years, and I learned a bunch in the process, but I would not want to drive this initiative. As to what I've learned (which makes me able to identify a list of specific open technical questions for which I see you have no answer yet), I can share that with you, but it certainly doesn't fit the format of a github comment.

**Comment by petermattis** [comment]
It is fun to be thinking about compiling expressions to native code, but I don't think we're ready for it yet. I think it is excellent to be experimenting now, but we need more experimentation before committing to a direction. We need evidence for the benefits and to balance those benefits against the concomitant burdens.

**Comment by petermattis** [comment]
In addition to pushing filters down into scan ops, we should also push projections so that we only return the needed columns. In conjunction with this we'll probably want to add another `ScanResponse` format that is composed of SQL rows and columns instead of keys and values.
